File: ios/DailyMemory/Domain/UseCase/Reminder/SnoozeReminderUseCase.swift
import Foundation

/// Use case for snoozing a reminder.
/// Creates a new scheduled time by adding the specified minutes.
final class SnoozeReminderUseCase {
    private let reminderRepository: ReminderRepository

    init(reminderRepository: ReminderRepository = ReminderRepositoryImpl()) {
        self.reminderRepository = reminderRepository
    }

    func execute(reminderId: String, snoozeMinutes: Int = 30) async throws {
        guard let reminder = try await reminderRepository.getById(reminderId) else {
            throw SnoozeError.reminderNotFound
        }

        let newScheduledTime = Calendar.current.date(
            byAdding: .minute,
            value: snoozeMinutes,
            to: Date()
        )!

        var updatedReminder = reminder
        updatedReminder.scheduledAt = newScheduledTime
        updatedReminder.triggeredAt = nil

        try await reminderRepository.update(updatedReminder)
    }

    enum SnoozeError: Error {
        case reminderNotFound
    }
}

File: android/app/src/main/java/com/dailymemory/domain/usecase/reminder/SnoozeReminderUseCase.kt
package com.dailymemory.domain.usecase.reminder

import com.dailymemory.domain.repository.ReminderRepository
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import javax.inject.Inject

/**
 * Use case for snoozing a reminder.
 * Creates a new scheduled time by adding the specified minutes.
 */
class SnoozeReminderUseCase @Inject constructor(
    private val reminderRepository: ReminderRepository
) {
    suspend operator fun invoke(reminderId: String, snoozeMinutes: Int = 30): Result<Unit> {
        return try {
            val reminder = reminderRepository.getById(reminderId)
                ?: return Result.failure(IllegalArgumentException("Reminder not found"))

            val newScheduledTime = LocalDateTime.now().plus(snoozeMinutes.toLong(), ChronoUnit.MINUTES)

            reminderRepository.update(
                reminder.copy(
                    scheduledAt = newScheduledTime,
                    triggeredAt = null
                )
            )

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
